### PR TITLE
One stable version.

### PR DESCRIPTION
**Pull request by MoonWalker1997**
I have optimized the code, make it better understanding, easier reading, and I found that there are no fatal and apparent bugs.

### CODE REVIEW

**Comment by bowen-xu** [comment]
I will do the code review as soon as possible.

File: (2).idea/PyNARS-0_0_2_pre_alpha.iml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bowen-xu) [inline (2).idea/PyNARS-0_0_2_pre_alpha.iml:1]: This file had better been contained in `.gitignore`.

File: (2).idea/misc.xml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bowen-xu) [inline (2).idea/misc.xml:1]: The folder had better been contained in `.gitignore`.

File: .idea/workspace.xml [hunk 3]
```diff
@@ -0,0 +1,185 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bowen-xu) [inline .idea/workspace.xml:1]: The folder had better been contained in `.gitignore`.

File: draft.py [hunk 4]
```diff
@@ -0,0 +1 @@
+import pynars.NARS.DataStructures.MC.SlotMC as SlotMC
```
Comment (bowen-xu) [inline draft.py:1]: This file should be deleted.

File: pynars/NARS/DataStructures/MC/AnticipationMC.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+from pynars.Narsese import Task, parser
```
Comment (bowen-xu) [inline pynars/NARS/DataStructures/MC/AnticipationMC.py:1]: An application is not a "DataStruction". We can create a new repo to develop GUIs.

File: pynars/NARS/DataStructures/MC/SampleChannels/NarseseChannel.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    def __init__(self, ID: str, num_slot, num_event, num_anticipation, num_prediction, memory):
+        super(NarseseChannel, self).__init__(ID, num_slot, num_event, num_anticipation, num_prediction, memory)
+        self.operations = [Term("^write")]
+        self.read_cursor = 0
+        self.count = 0
+
+
+    def execute(self, term: Term):
+        if term.equal(self.operations[0]):
+            f = open("C://Users//TORY//Downloads//PyNARS-0_0_2_pre_alpha ("
```
Comment (bowen-xu) [inline pynars/NARS/DataStructures/MC/SampleChannels/NarseseChannel.py:16]: The file path is the absolute one. It's valid on your computer, however, it may not correct in other's computers. So the relative path is better, or you can get the absolute path dynamically by using `pathlib` or `os` package.

File: pynars/NARS/DataStructures/MC/SampleChannels/NarseseChannel.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                     "1)//PyNARS-0_0_2_pre_alpha//pynars//NARS//DataStructures//MC//playground.txt", "w")
+            f.write("I am writing" + str(self.count))
+            self.count += 1
+            f.close()
+
+    def information_gathering(self):
+        f = open("C://Users//TORY//Downloads//PyNARS-0_0_2_pre_alpha ("
```
Comment (bowen-xu) [inline pynars/NARS/DataStructures/MC/SampleChannels/NarseseChannel.py:23]: The same  issue as the above.

File: pynars/NARS/DataStructures/MC/SampleChannels/SampleChannel1.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# from pynars.NARS.DataStructures.MC.ChannelMC import ChannelMC
```
Comment (bowen-xu) [inline pynars/NARS/DataStructures/MC/SampleChannels/SampleChannel1.py:1]: This file only cotains comments wihout useful code? If so, it should be deleted.

File: pynars/NARS/DataStructures/MC/SampleChannels/SampleEnvironment1.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
+import numpy as np
```
Comment (bowen-xu) [inline pynars/NARS/DataStructures/MC/SampleChannels/SampleEnvironment1.py:1]: Is the "SampleEnvironment*" a kind of environment insead of a Channel? If so, they had better been moved into a seprated folder out of the `MC` folder, or we can create a new repo to do the experiments.

File: pynars/NARS/DataStructures/MC/SlotMC_archieve.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
+import numpy as np
```
Comment (bowen-xu) [inline pynars/NARS/DataStructures/MC/SlotMC_archieve.py:1]: Is this file deprecated? If so, it should not be submitted.

File: pynars/NARS/DataStructures/_py/Bag.py [hunk 11]
```diff
@@ -7,32 +7,32 @@
...
 
+
 class Bag:
     # TODO: Re-implement this DataStructure, in order to optimize the complexity.
     class LUT:
         def __init__(self, *args, **kwargs):
             self.lut = OrderedDict(*args, **kwargs)
 
-        def get(self, key, default=None):
+        def get(self, key, default = None):
```
Comment (bowen-xu) [inline pynars/NARS/DataStructures/_py/Bag.py:17]: `default=None` looks better. Maybe we can discuss what style to choose as our standard.

File: pynars/NARS/DataStructures/MC/playground.txt [hunk 12]
```diff
@@ -0,0 +1,36 @@
+1
```
Comment (bowen-xu) [inline pynars/NARS/DataStructures/MC/playground.txt:1]: It looks like a part of the "environment"?

File: pynars/NARS/DataStructures/_py/Table.py [hunk 13]
```diff
@@ -10,8 +10,11 @@ def __init__(self, capacity):
         self._table = DEPQ(maxlen=capacity)
 
     def add(self, task: Task, p: float):
+        if task in self:
```
Comment (bowen-xu) [inline pynars/NARS/DataStructures/_py/Table.py:13]: Why remove before insert?

**Comment by bowen-xu** [comment]
Is the file `pynars/moon.jpg` necessary to upload?

**Review by bowen-xu — APPROVED** [review]

File: pynars/NARS/DataStructures/_py/Bag.py (hunk 11 above)
Comment (MoonWalker1997) [reply pynars/NARS/DataStructures/_py/Bag.py:17]: I usually use Pycharm's "automatic reformatting", it "sets" the default formatting this way. But the way, I think it is necessary to set up a "naming policy". For me, I prefers double-quotes ("") better than single-quote (''). And I would like to set long comments (""" """) under a function when this comment is related to it directly. And all default values with spaces (default = None, than default=None) in function definition. But in function calling, I will skip this space. 

In short, I can set up a Pycharm formatting file to regularize this standard.

File: pynars/NARS/DataStructures/MC/SampleChannels/SampleEnvironment1.py (hunk 9 above)
Comment (MoonWalker1997) [reply pynars/NARS/DataStructures/MC/SampleChannels/SampleEnvironment1.py:1]: Yes. But in this version, the environment and NARS is hard bounded. We may allow channels detach from the NARS core, but it is not in this version. So, I just place these channels and we don't have a "pure software" as a result. But this will be solved in the next version (UI updating version).

File: pynars/NARS/DataStructures/MC/SampleChannels/SampleChannel1.py (hunk 8 above)
Comment (MoonWalker1997) [reply pynars/NARS/DataStructures/MC/SampleChannels/SampleChannel1.py]: I have deleted. Leaving it as it be was because deleting it will let the sample channels start at 2, which is not very natural.

File: pynars/NARS/DataStructures/MC/SampleChannels/NarseseChannel.py (hunk 6 above)
Comment (MoonWalker1997) [reply pynars/NARS/DataStructures/MC/SampleChannels/NarseseChannel.py]: I have changed it. And I have also moved the "playground.txt" file to the same working directory.

File: pynars/NARS/DataStructures/MC/playground.txt (hunk 12 above)
Comment (MoonWalker1997) [reply pynars/NARS/DataStructures/MC/playground.txt]: Yes. This is a very simple simulation, it asks the agent to write something or read from the txt file, and so that is why it is called a "playground".
